clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -57,238 +55,51 @@
-                             List<AbortedTransaction> abortedTransactions,
-                             T records) {
-            this(error, highWatermark, lastStableOffset, logStartOffset, Optional.empty(), abortedTransactions, records);
-        }
-
-        @Override
-        public boolean equals(Object o) {
-            if (this == o)
-                return true;
-            if (o == null || getClass() != o.getClass())
-                return false;
-
-            PartitionData that = (PartitionData) o;
-
-            return this.partitionResponse.equals(that.partitionResponse);
-        }
-
-        @Override
-        public int hashCode() {
-            return this.partitionResponse.hashCode();
-        }
-
-        @Override
-        public String toString() {
-            return "(error=" + error() +
-                    ", highWaterMark=" + highWatermark() +
-                    ", lastStableOffset = " + lastStableOffset() +
-                    ", logStartOffset = " + logStartOffset() +
-                    ", preferredReadReplica = " + preferredReadReplica().map(Object::toString).orElse("absent") +
-                    ", abortedTransactions = " + abortedTransactions() +
-                    ", divergingEpoch =" + divergingEpoch() +
-                    ", recordsSizeInBytes=" + records().sizeInBytes() + ")";
-        }
-
-        public Errors error() {
-            return error;
-        }
-
-        public long highWatermark() {
-            return partitionResponse.highWatermark();
-        }
-
-        public long lastStableOffset() {
-            return partitionResponse.lastStableOffset();
-        }
-
-        public long logStartOffset() {
-            return partitionResponse.logStartOffset();
-        }
-
-        public Optional<Integer> preferredReadReplica() {
-            return preferredReplica;
-        }
-
-        public List<AbortedTransaction> abortedTransactions() {
-            return abortedTransactions;
-        }
-
-        public Optional<FetchResponseData.EpochEndOffset> divergingEpoch() {
-            FetchResponseData.EpochEndOffset epochEndOffset = partitionResponse.divergingEpoch();
-            if (epochEndOffset.epoch() < 0) {
-                return Optional.empty();
-            } else {
-                return Optional.of(epochEndOffset);
-            }
-        }
-
-        @SuppressWarnings("unchecked")
-        public T records() {
-            return (T) partitionResponse.recordSet();
-        }
-    }
-
-    /**
-     * From version 3 or later, the entries in `responseData` should be in the same order as the entries in
-     * `FetchRequest.fetchData`.
-     *
-     * @param error             The top-level error code.
-     * @param responseData      The fetched data grouped by partition.
-     * @param throttleTimeMs    The time in milliseconds that the response was throttled
-     * @param sessionId         The fetch session id.
-     */
     public FetchResponse(Errors error,
```

Comment:
I think it would be better to make this a static factory method and keep the constructor for the case where we receive `FetchResponseData`.